File: src/crypto.c
/*****************************************************************************
 *   Ledger Stellar App.
 *   (c) 2024 Ledger SAS.
 *   (c) 2024 overcat.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *****************************************************************************/

#include <stdint.h>  // uint*_t
#include <string.h>  // explicit_bzero

#include "crypto.h"
#include "globals.h"

#define STELLAR_SEED_KEY "ed25519 seed"

static cx_err_t crypto_derive_private_key(cx_ecfp_private_key_t *private_key,
                                          const uint32_t *bip32_path,
                                          uint8_t bip32_path_len) {
    uint8_t raw_private_key[64] = {0};
    cx_err_t error = CX_OK;

    // derive the seed with bip32_path
    CX_CHECK(os_derive_bip32_with_seed_no_throw(HDW_ED25519_SLIP10,
                                                CX_CURVE_Ed25519,
                                                bip32_path,
                                                bip32_path_len,
                                                raw_private_key,
                                                NULL,
                                                (unsigned char *) STELLAR_SEED_KEY,
                                                sizeof(STELLAR_SEED_KEY)));
    // new private_key from raw
    CX_CHECK(cx_ecfp_init_private_key_no_throw(CX_CURVE_Ed25519,
                                               raw_private_key,
                                               RAW_ED25519_PRIVATE_KEY_SIZE,
                                               private_key));

end:
    explicit_bzero(&raw_private_key, sizeof(raw_private_key));
    if (error != CX_OK) {
        explicit_bzero(private_key, sizeof(*private_key));
        PRINTF("In crypto_derive_private_key: ERROR %x \n", error);
    }
    return error;
}

// converts little endian 32 byte public key to big endian 32 byte public key
static void raw_public_key_le_to_be(cx_ecfp_public_key_t *public_key,
                                    uint8_t raw_public_key[static RAW_ED25519_PUBLIC_KEY_SIZE]) {
    // copy public key little endian to big endian
    for (uint8_t i = 0; i < RAW_ED25519_PUBLIC_KEY_SIZE; i++) {
        raw_public_key[i] = public_key->W[64 - i];
    }
    // set sign bit
    if ((public_key->W[32] & 1) != 0) {
        raw_public_key[31] |= 0x80;
    }
}

cx_err_t crypto_derive_public_key(uint8_t raw_public_key[static RAW_ED25519_PUBLIC_KEY_SIZE],
                                  const uint32_t *bip32_path,
                                  uint8_t bip32_path_len) {
    cx_err_t error = CX_OK;
    cx_ecfp_private_key_t private_key = {0};
    cx_ecfp_public_key_t public_key = {0};

    // derive private key according to BIP32 path
    CX_CHECK(crypto_derive_private_key(&private_key, bip32_path, bip32_path_len));

    // generate corresponding public key
    CX_CHECK(cx_ecfp_generate_pair_no_throw(CX_CURVE_Ed25519, &public_key, &private_key, 1));

    raw_public_key_le_to_be(&public_key, raw_public_key);

end:
    explicit_bzero(&private_key, sizeof(private_key));
    if (error != CX_OK) {
        PRINTF("In crypto_init_public_key: ERROR %x \n", error);
    }
    return error;
}

cx_err_t crypto_sign_message(const uint8_t *message,
                             uint8_t message_len,
                             const uint8_t *signature,
                             uint8_t signature_len,
                             const uint32_t *bip32_path,
                             uint8_t bip32_path_len) {
    cx_err_t error = CX_OK;
    cx_ecfp_private_key_t private_key = {0};

    // derive private key according to BIP32 path
    CX_CHECK(crypto_derive_private_key(&private_key, bip32_path, bip32_path_len));

    CX_CHECK(cx_eddsa_sign_no_throw(&private_key,
                                    CX_SHA512,
                                    message,
                                    message_len,
                                    (unsigned char *) signature,
                                    signature_len));
    PRINTF("Signature: %.*H\n", signature_len, signature);

end:
    explicit_bzero(&private_key, sizeof(private_key));
    if (error != CX_OK) {
        PRINTF("In crypto_sign_message: ERROR %x \n", error);
    }
    return error;
}

File: src/crypto.rs
/*****************************************************************************
 *   Ledger App Stellar Rust.
 *   (c) 2025 overcat
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *****************************************************************************/

use crate::{bip32::Bip32Path, sw::AppSW};
use ledger_device_sdk::{
    ecc::{ECPublicKey, Ed25519},
    hash::{sha2::Sha2_256, HashInit},
};

const STELLAR_COMPRESSED_PK_LEN: usize = 32;
const LEDGER_UNCOMPRESSED_PK_LEN: usize = 65;
const HASH_LEN: usize = 32;
const SIGNATURE_LEN: usize = 64;

/// Derives the Ed25519 public key for a given BIP32 path and converts it to Stellar format.
///
/// This function derives a private key using SLIP-10 derivation from the provided BIP32 path,
/// computes its corresponding public key, and converts it to Stellar's 32-byte compressed format.
///
/// # Arguments
///
/// * `path` - The BIP32 derivation path (typically 44'/148'/account' for Stellar)
///
/// # Returns
///
/// * `Ok([u8; 32])` - The 32-byte Stellar-formatted public key on success
/// * `Err(AppSW::KeyDeriveFail)` - If key derivation fails
///
/// # Security
///
/// The private key is derived from the device's secure element and is never stored in memory.
/// It exists only temporarily during the execution of this function.
pub fn get_public_key(path: &Bip32Path) -> Result<[u8; STELLAR_COMPRESSED_PK_LEN], AppSW> {
    let private_key = Ed25519::derive_from_path_slip10(path.as_ref());
    let public_key = private_key.public_key().map_err(|_| AppSW::KeyDeriveFail)?;
    Ok(convert_to_stellar_pk(&public_key))
}

/// Computes the SHA-256 hash of the provided data.
///
/// This function uses the hardware-accelerated SHA-256 implementation from the Ledger SDK
/// to compute a cryptographic hash of the input data.
///
/// # Arguments
///
/// * `data` - The byte slice to hash
///
/// # Returns
///
/// * `Ok([u8; 32])` - The 32-byte SHA-256 hash on success
/// * `Err(AppSW::DataHashFail)` - If the hashing operation fails
///
/// # Example Use Cases
///
/// - Hashing transaction data before signing
/// - Computing message digests for signature verification
pub fn hash(data: &[u8]) -> Result<[u8; HASH_LEN], AppSW> {
    let mut hash: [u8; HASH_LEN] = [0u8; HASH_LEN];
    Sha2_256::new()
        .hash(data, &mut hash)
        .map_err(|_| AppSW::DataHashFail)?;
    Ok(hash)
}

/// Signs data using Ed25519 with a key derived from the given BIP32 path.
///
/// This function derives a private key using SLIP-10 derivation from the provided BIP32 path
/// and uses it to create an Ed25519 signature over the input data (typically a hash).
///
/// # Arguments
///
/// * `data` - The data to sign (typically a 32-byte hash of the actual message/transaction)
/// * `path` - The BIP32 derivation path (typically 44'/148'/account' for Stellar)
///
/// # Returns
///
/// * `Ok([u8; 64])` - The 64-byte Ed25519 signature on success
/// * `Err(AppSW::DataSignFail)` - If the signing operation fails
pub fn sign(data: &[u8], path: &Bip32Path) -> Result<[u8; SIGNATURE_LEN], AppSW> {
    let private_key = Ed25519::derive_from_path_slip10(path.as_ref());
    let (signature, _) = private_key.sign(data).map_err(|_| AppSW::DataSignFail)?;
    Ok(signature)
}

/// Converts an Ed25519 public key from Ledger's uncompressed format to Stellar's compressed format.
///
/// # Arguments
///
/// * `input` - A 65-byte Ed25519 public key in big-endian uncompressed format from the Ledger SDK
///
/// # Returns
///
/// * `[u8; 32]` - The 32-byte Stellar-formatted public key
pub fn convert_to_stellar_pk(
    input: &ECPublicKey<LEDGER_UNCOMPRESSED_PK_LEN, 'E'>,
) -> [u8; STELLAR_COMPRESSED_PK_LEN] {
    let mut out = [0u8; STELLAR_COMPRESSED_PK_LEN];

    for (i, out_byte) in out.iter_mut().enumerate() {
        *out_byte = input.pubkey[LEDGER_UNCOMPRESSED_PK_LEN - 1 - i];
    }

    if (input.pubkey[STELLAR_COMPRESSED_PK_LEN] & 1) != 0 {
        out[STELLAR_COMPRESSED_PK_LEN - 1] |= 0x80;
    }

    out
}
